In [38]:
import numpy as np
import pyreadr
import statsmodels.api as sm

# Exercise 3.1

In [15]:
catheter_r = pyreadr.read_r('../Data/catheter.rda')
catheter_df = catheter_r['catheter']
catheter_df

height  weight  catlength
0    107.4    18.0       37.0
1    159.4    42.1       49.5
2     94.1    16.0       34.5
3     99.1    13.5       36.0
4    114.2    23.4       43.0
5     96.6     7.7       28.0
6    107.9    17.3       37.0
7     56.5     3.8       20.0
8     92.9    14.9       33.5
9     59.0     4.3       30.5
10    82.8     9.5       38.5
11   145.6    35.6       47.0

a)

In [16]:
x_height = sm.add_constant(catheter_df.height)
y = catheter_df.catlength

modelled_by_height = sm.OLS(y, x_height).fit()

modelled_by_height.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:              catlength   R-squared:                       0.776
Model:                            OLS   Adj. R-squared:                  0.754
Method:                 Least Squares   F-statistic:                     34.73
Date:                Mon, 06 Oct 2025   Prob (F-statistic):           0.000152
Time:                        15:25:58   Log-Likelihood:                -32.594
No. Observations:                  12   AIC:                             69.19
Df Residuals:                      10   BIC:                             70.16
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         12.1271      4.247      2.855      0.017       2.664      21.590
height         0.2377      0.040      5.893      0.000       0.148       0.328
==============================================================================
Omnibus:                        0.408   Durbin-Watson:                   1.790
Prob(Omnibus):                  0.815   Jarque-Bera (JB):                0.073
Skew:                          -0.165   Prob(JB):                        0.964
Kurtosis:                       2.806   Cond. No.                         386.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [17]:
x_weight = sm.add_constant(catheter_df.weight)
y = catheter_df.catlength

modelled_by_weight = sm.OLS(y, x_weight).fit()

modelled_by_weight.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:              catlength   R-squared:                       0.799
Model:                            OLS   Adj. R-squared:                  0.779
Method:                 Least Squares   F-statistic:                     39.86
Date:                Mon, 06 Oct 2025   Prob (F-statistic):           8.75e-05
Time:                        15:25:59   Log-Likelihood:                -31.943
No. Observations:                  12   AIC:                             67.89
Df Residuals:                      10   BIC:                             68.86
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         25.6263      2.003     12.796      0.000      21.164      30.088
weight         0.6161      0.098      6.313      0.000       0.399       0.834
==============================================================================
Omnibus:                        2.618   Durbin-Watson:                   1.854
Prob(Omnibus):                  0.270   Jarque-Bera (JB):                0.500
Skew:                          -0.279   Prob(JB):                        0.779
Kurtosis:                       3.830   Cond. No.                         37.6
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

**Predictor Significance**

**For `modelled_by_height`:**
- **Height predictor**: p-value = 0.000 (< 0.05) → **Significant** ✓
- The height coefficient is 0.2377 with P>|t| = 0.000

**For `modelled_by_weight`:**
- **Weight predictor**: p-value = 0.000 (< 0.05) → **Significant** ✓
- The weight coefficient is 0.6161 with P>|t| = 0.000

Both predictors are highly significant in their respective models. Both p-values are well below the conventional significance level of 0.05, indicating strong evidence that both height and weight have a significant relationship with catheter length.

Additionally:
- The F-statistic for both models also shows very low p-values (0.000152 for height, 8.75e-05 for weight), confirming overall model significance
- Weight model has slightly higher R-squared (0.799 vs 0.776), suggesting it explains slightly more variance in catheter length than height alone

b)

In [22]:
x_height_and_weight = sm.add_constant(catheter_df[['height', 'weight']])
y = catheter_df.catlength

modelled_by_height_and_weight = sm.OLS(y, x_height_and_weight).fit()
modelled_by_height_and_weight.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:              catlength   R-squared:                       0.806
Model:                            OLS   Adj. R-squared:                  0.762
Method:                 Least Squares   F-statistic:                     18.65
Date:                Mon, 06 Oct 2025   Prob (F-statistic):           0.000630
Time:                        15:33:20   Log-Likelihood:                -31.757
No. Observations:                  12   AIC:                             69.51
Df Residuals:                       9   BIC:                             70.97
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         21.0853      8.770      2.404      0.040       1.245      40.925
height         0.0768      0.144      0.533      0.607      -0.249       0.403
weight         0.4275      0.368      1.161      0.275      -0.405       1.260
==============================================================================
Omnibus:                        1.049   Durbin-Watson:                   1.814
Prob(Omnibus):                  0.592   Jarque-Bera (JB):                0.029
Skew:                          -0.017   Prob(JB):                        0.986
Kurtosis:                       3.239   Cond. No.                         826.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

**Predictor Significance in Multiple Linear Regression**

**Individual Predictor Significance:**

**Height:**
- Coefficient: 0.0768
- p-value: 0.607 (> 0.05) → **NOT Significant** ✗
- 95% CI: [-0.249, 0.403] (includes zero)

**Weight:**
- Coefficient: 0.4275
- p-value: 0.275 (> 0.05) → **NOT Significant** ✗
- 95% CI: [-0.405, 1.260] (includes zero)

**Overall Model Significance:**

- **F-statistic**: 18.65 with p-value (= "Prob (F-statistic)") = 0.000630 (< 0.05) → **Model is Significant** ✓
- **R-squared**: 0.806 (80.6% of variance explained)
- **Adj. R-squared**: 0.762

**Key Interpretation:**

This is a classic case of **multicollinearity**!

**What's happening:**
1. **Overall model is significant** - height and weight together significantly predict catheter length
2. **Individual predictors are NOT significant** - neither appears significant when controlling for the other
3. **High condition number (826)** - indicates multicollinearity between height and weight

**Why this occurs:**

- Height and weight are highly correlated with each other
- When both are in the model, they "compete" to explain the same variance
- Each predictor's unique contribution (after accounting for the other) is not statistically significant
- Yet together, they explain substantial variance (R² = 0.806)

**Comparison Summary:**

| Model | Predictor(s) Significant? | R² |
| --- | --- | --- |
| Height only | ✓ Yes (p < 0.001) | 0.776 |
| Weight only | ✓ Yes (p < 0.001) | 0.799 |
| Height + Weight (MLR) | ✗ Neither (p > 0.05) | 0.806 |

**Conclusion:** The predictors are NOT individually significant in the multiple linear regression, likely due to multicollinearity between height and weight.


c) As seen in the summary output above, both predictors are insignificant:

- height: p-value = 0.697 which is not < 0.05
- weight: p-value = 0.275 which is not < 0.05

Is this a contradiction to the results obtained in the two simple linear regression models? The answer is no. In multiple linear regression, on the basis of the hypotheses tests we decide whether the predictor variable height is required when the value of the predictor weight is known. The answer is no and the same holds vice versa.

However, the gllobal F-test indicates that we need at least one of the two predictors:

- Prob (F-statistic) = 0.00063 which is < 0.05

**Conclusion:**

So we do not need to include both predictors simultaneously but
we need one of them. This situation occurs when the predictors are strongly
correlated. Due to the smaller p-value (= 0.275) we would prefer to include the predictor
weight in the regression model.